src/clients/kol.ts

```diff
@@ -87,95 +114,239 @@ function sanitiseBlueText(blueText: string | undefined): string {
+      } catch {
+        console.log("Login failed. Checking if it's because of rollover.");
+        await this.rolloverCheck();
+        return false;
       }
     });
   }
 
-  private async makeCredentialedRequest(url: string, parameters: object) {
+  private chatBotStarted = false;
+
+  async startChatBot() {
+    if (this.chatBotStarted) return;
+    this.loopChatBot();
+    this.chatBotStarted = true;
+  }
+
+  private async loopChatBot() {
+    await Promise.all([this.checkWhispers(), this.checkKmails()]);
+    await wait(3000);
+    await this.loopChatBot();
+  }
+
+  private lastFetchedWhispers = "0";
+
+  async checkWhispers() {
+    const newChatMessagesResponse = (await this.visitUrl("newchatmessages.php", {
+      j: 1,
+      lasttime: this.lastFetchedWhispers,
+    })) as { last: string; msgs: KoLChatMessage[] };
+
+    if (!newChatMessagesResponse || typeof newChatMessagesResponse !== "object") return;
+
+    this.lastFetchedWhispers = newChatMessagesResponse["last"];
+
+    newChatMessagesResponse["msgs"]
+      .filter((msg) => msg["type"] === "private" && msg.who && msg.msg)
+      .map((msg) => ({
+        who: {
+          id: Number(msg.who!.id),
+          name: msg.who!.name,
+        },
+        msg: msg.msg!,
+        time: new Date(Number(msg.time) * 1000),
+      }))
+      .forEach((whisper) => this.emit("whisper", whisper));
+  }
+
+  async checkKmails() {
+    const newKmailsResponse = (await this.visitUrl("api.php", {
+      what: "kmail",
+      for: `${this._loginParameters.get("loginname")} Chatbot`,
+    })) as KoLKmail[];
+
+    if (!Array.isArray(newKmailsResponse) || newKmailsResponse.length === 0) return;
+
+    const newKmails = newKmailsResponse.map((msg: KoLKmail) => ({
+      who: {
+        id: Number(msg.fromid),
+        name: msg.fromname,
+      },
+      msg: msg.message,
+      time: new Date(Number(msg.azunixtime) * 1000),
+    }));
+
+    const data = {
+      the_action: "delete",
+      pwd: this._credentials?.pwdhash,
+      box: "Inbox",
+      ...newKmailsResponse.reduce(
+        (acc: { [x: string]: string }, msg: KoLKmail) => ({
+          ...acc,
+          [`sel${msg.id}`]: "on",
+        }),
+        {}
+      ),
+    };
+
+    await this.visitUrl("messages.php", data);
+
+    newKmails.forEach((m) => this.emit("kmail", m));
+  }
+
+  async useChatMacro(macro: string) {
+    await this.visitUrl("submitnewchat.php", {
+      graf: `/clan ${macro}`,
+      j: 1,
+    });
+  }
+
+  async whisper(recipientId: number, message: string) {
+    await this.useChatMacro(`/w ${recipientId} ${message}`);
+  }
+
+  async rolloverCheck() {
+    this._isRollover = /The system is currently down for nightly maintenance/.test(
+      (await axios("https://www.kingdomofloathing.com/")).data
+    );
+    if (this._isRollover) {
+      console.log("Rollover appears to be in progress. Checking again in one minute.");
+      setTimeout(() => this.rolloverCheck(), 60000);
+    }
+  }
+
+  async visitUrl(
```

Comment:
I think this is a pithier and probably better name than `tryRequestWithLogin`, but it does elide the side effect of logging in in a way that may be counterintuitive to other developers.

Reply:
The reality is that the old way just didn't work in all circumstances and you could get into weird race conditions of logging in. This just works which rocks

Reply:
oh, I mean, I think the change is great. I just think that the name `visitUrl` might be too generic and disguise the fact that it'll also log in if necessary. I don't think it's that big a deal, but it's worth noting here.